client-tez/src/main/java/org/apache/tez/runtime/library/common/shuffle/impl/RssSimpleFetchedInputAllocator.java

```diff
@@ -0,0 +1,201 @@
+    final float maxInMemCopyUse = conf.getFloat(
+        TezRuntimeConfiguration.TEZ_RUNTIME_SHUFFLE_FETCH_BUFFER_PERCENT,
+        TezRuntimeConfiguration.TEZ_RUNTIME_SHUFFLE_FETCH_BUFFER_PERCENT_DEFAULT);
+    if (maxInMemCopyUse > 1.0 || maxInMemCopyUse < 0.0) {
+      throw new IllegalArgumentException("Invalid value for "
+          + TezRuntimeConfiguration.TEZ_RUNTIME_SHUFFLE_FETCH_BUFFER_PERCENT + ": " + maxInMemCopyUse);
+    }
+    return (long) (conf.getLong(Constants.TEZ_RUNTIME_TASK_MEMORY,
+        Math.min(maxAvailableTaskMemory, Integer.MAX_VALUE)) * maxInMemCopyUse);
+  }
+
+  @Override
+  public synchronized FetchedInput allocate(long actualSize, long compressedSize,
+            InputAttemptIdentifier inputAttemptIdentifier) throws IOException {
+    if (actualSize > maxSingleShuffleLimit
+        || this.usedMemory + actualSize > this.memoryLimit) {
+      LOG.info("Allocate DiskFetchedInput, length:{}", actualSize);
+      return new DiskFetchedInput(actualSize + 8, inputAttemptIdentifier, this, conf,
+          localDirAllocator, fileNameAllocator);
+    } else {
+      this.usedMemory += actualSize;
+      if (LOG.isDebugEnabled()) {
+        LOG.info(srcNameTrimmed + ": " + "Used memory after allocating " + actualSize + " : " + usedMemory);
+      }
+      return new MemoryFetchedInput(actualSize, inputAttemptIdentifier, this);
+    }
+  }
+
+  @Override
+  public synchronized FetchedInput allocateType(Type type, long actualSize,
+      long compressedSize, InputAttemptIdentifier inputAttemptIdentifier)
+      throws IOException {
+
+    switch (type) {
+      case DISK:
+        LOG.info("AllocateType DiskFetchedInput, compressedSize:{}", compressedSize);
+        return new DiskFetchedInput(compressedSize + 8, inputAttemptIdentifier, this,
+            conf, localDirAllocator, fileNameAllocator);
+      default:
+        return allocate(actualSize, compressedSize, inputAttemptIdentifier);
+    }
+  }
+
+  @Override
+  public synchronized void fetchComplete(FetchedInput fetchedInput) {
+    switch (fetchedInput.getType()) {
+      // Not tracking anything here.
+      case DISK:
+      case DISK_DIRECT:
+      case MEMORY:
+        break;
+      default:
+        throw new TezUncheckedException("InputType: " + fetchedInput.getType() + " not expected for Broadcast fetch");
+    }
+  }
+
+  @Override
+  public synchronized void fetchFailed(FetchedInput fetchedInput) {
+    cleanup(fetchedInput);
+  }
+
+  @Override
+  public synchronized void freeResources(FetchedInput fetchedInput) {
+    cleanup(fetchedInput);
+  }
+
+  private void cleanup(FetchedInput fetchedInput) {
+    switch (fetchedInput.getType()) {
+      case DISK:
+        break;
+      case MEMORY:
+        unreserve(((MemoryFetchedInput) fetchedInput).getSize());
+        break;
+      default:
+        throw new TezUncheckedException("InputType: " + fetchedInput.getType() + " not expected for Broadcast fetch");
```

Comment:
Could we use RssException? Because we use count the RssException to calculate the fail jobs caused by rss.

Reply:
ok, we modified it